cli_tools/gce_windows_upgrade/upgrader/upgrade_paths.go

```diff
@@ -0,0 +1,126 @@
+	licenseForWindows2022Upgraded   = "projects/windows-cloud/global/licenses/windows-server-2022-dc-in-place-upgrade"
+
+	versionStringForWindows2008r2 = "Windows Server 2008 R2 Datacenter"
+	versionStringForWindows2012r2 = "Windows Server 2012 R2 Datacenter"
+	versionStringForWindows2016   = "Windows Server 2016 Datacenter"
+	versionStringForWindows2019   = "Windows Server 2019 Datacenter"
+	versionStringForWindows2022   = "Windows Server 2022 Datacenter"
+)
+
+// SupportedVersions shows all supported OS versions of the tool.
+var (
+	SupportedVersions = []string{
+		versionWindows2008r2,
+		versionWindows2012r2,
+		versionWindows2016,
+		versionWindows2019,
+		versionWindows2022,
+	}
+
+	expectedCurrentLicenseForSourceOS = map[string][]string{
+		versionWindows2008r2: {licenseForWindows2008r2},
+		versionWindows2012r2: {licenseForWindows2012r2, licenseForWindows2012r2Upgraded},
+		versionWindows2016:   {licenseForWindows2016, licenseForWindows2016Upgraded},
+		versionWindows2019:   {licenseForWindows2019, licenseForWindows2019Upgraded},
+	}
+
+	licenseToAddForTargetOS = map[string]string{
+		versionWindows2012r2: licenseForWindows2012r2Upgraded,
+		versionWindows2016:   licenseForWindows2016Upgraded,
+		versionWindows2019:   licenseForWindows2019Upgraded,
+		versionWindows2022:   licenseForWindows2022Upgraded,
+	}
+
+	versionString = map[string]string{
+		versionWindows2008r2: versionStringForWindows2008r2,
+		versionWindows2012r2: versionStringForWindows2012r2,
+		versionWindows2016:   versionStringForWindows2016,
+		versionWindows2019:   versionStringForWindows2019,
+		versionWindows2022:   versionStringForWindows2022,
+	}
+
+	installFolderForTargetOS = map[string]string{
+		versionWindows2012r2: "Windows_Svr_Std_and_DataCtr_2012_R2_64Bit_English",
+		versionWindows2016:   "Win_Server_STD_CORE_2016_64Bit_English",
+		versionWindows2019:   "Win_Server_STD_CORE_2019_1809.1_64Bit_English",
+		versionWindows2022:   "Win_Server_STD_CORE_2022_64Bit_English",
+	}
+)
+
+type upgradePath struct {
+	enabled                bool
+	expectedCurrentLicense []string
+	licenseToAdd           string
+	expectedCurrentVersion string
+	expectedNewVersion     string
+	installFolder          string
+}
+
+func initUpgradePath(u map[string]map[string]*upgradePath) map[string]map[string]*upgradePath {
+	for sourceOS, targets := range u {
+		for targetOS, upgradePath := range targets {
+			upgradePath.expectedCurrentLicense = expectedCurrentLicenseForSourceOS[sourceOS]
+			upgradePath.licenseToAdd = licenseToAddForTargetOS[targetOS]
+			upgradePath.expectedCurrentVersion = versionString[sourceOS]
+			upgradePath.expectedNewVersion = versionString[targetOS]
+			upgradePath.installFolder = installFolderForTargetOS[targetOS]
+		}
+	}
+	return u
+}
+
+var upgradePaths = initUpgradePath(map[string]map[string]*upgradePath{
+	versionWindows2008r2: {
+		versionWindows2012r2: {enabled: true},
+	},
+	versionWindows2012r2: {
```

Comment:
Doesn't "enabled: true" enabled upgrades for 2012r2/2016/2019?

Reply:
In the previous commit, they should be disabled; now they all should be enabled for the e2e purpose. The gcloud gated them for the real customers, so it's not a problem to enable them for e2e.